app/controllers/build_estimates_controller.rb

```diff
@@ -1,61 +1,70 @@
 class BuildEstimatesController < ApplicationController
   include Wicked::Wizard
-  include BuildEstimatesHelper
+  include StepsHelper
 
-  steps(*ALL_ESTIMATE_STEPS)
+  steps(*STEPS_WITH_PROPERTY)
 
-  MONTHLY_INCOME_ATTRIBUTES = MonthlyIncomeForm::INCOME_ATTRIBUTES + [:monthly_incomes]
-
-  def new
-    session[session_key] = {}
-    redirect_to wizard_path(steps.first)
-  end
+  HANDLER_CLASSES = {
+    applicant: Flow::ApplicantHandler,
+    monthly_income: Flow::IncomeHandler,
+    property: Flow::PropertyHandler,
+    vehicle: Flow::VehicleHandler,
+    assets: Flow::AssetHandler,
+    summary: Flow::SummaryHandler,
+    outgoings: Flow::OutgoingsHandler,
+    property_entry: Flow::PropertyEntryHandler,
+  }.freeze
 
   def show
-    case step
-    when :intro
-      @estimate = IntroForm.new session[session_key]
-    when :monthly_income
-      @estimate = MonthlyIncomeForm.new session[session_key].slice(*MONTHLY_INCOME_ATTRIBUTES)
-    when :property
-      @intro_form = IntroForm.new session[session_key]
-    end
+    handler = HANDLER_CLASSES[step]
+    @form = if handler
+              handler.model(session_data)
+            else
+              cfe_connection.api_result(estimate_id)
```

Comment:
It might be that we want to display a page with CFE's calculation results more than once (e.g. if the user views the results page and then the print-friendly result page). Given that, rather than call CFE at the point where we first want to display the results, would it make more sense to call it when we submit the penultimate page, and stash the result in the session? Then we could add a Flow::CalculationResultHandler and do away with this if/else statement entirely.

If that's beyond the scope of this ticket then ignore me entirely!

Reply:
That would solve the problem. I'm not sure we want to cache the API result in the session though - there might be too many places to forget to update it. I wanted to push everything to the API, but it doesn't allow that (specifically w/o an applicant the 'GET' fails, and applicants can't be overwritten). Maybe we should consider 'fixing' that in our dev meeting - it is weird that we collect the applicant data first but submit it last